packages/graph/hash_graph/lib/graph/src/types/schema/data.rs

```diff
@@ -0,0 +1,316 @@
+    }
+
+    #[must_use]
+    pub fn ty(&self) -> &str {
+        &self.ty
+    }
+
+    #[must_use]
+    pub const fn additional_properties(&self) -> &HashMap<String, serde_json::Value> {
+        &self.additional_properties
+    }
+
+    #[must_use]
+    pub fn additional_properties_mut(&mut self) -> &mut HashMap<String, serde_json::Value> {
+        &mut self.additional_properties
+    }
+
+    /// Returns the predefined text data type.
+    #[must_use]
+    pub fn text() -> Self {
+        Self::new_unchecked(
+            Uri::new_unchecked("https://blockprotocol.org/types/@blockprotocol/data-type/text"),
+            "Text",
+            "An ordered sequence of characters".to_owned(),
+            "string",
+            HashMap::default(),
+        )
+    }
+
+    /// Returns the predefined number data type.
+    #[must_use]
+    pub fn number() -> Self {
+        Self::new_unchecked(
+            Uri::new_unchecked("https://blockprotocol.org/types/@blockprotocol/data-type/number"),
+            "Number",
+            "An arithmetical value (in the Real number system)".to_owned(),
+            "number",
+            HashMap::default(),
+        )
+    }
+
+    /// Returns the predefined boolean data type.
+    #[must_use]
+    pub fn boolean() -> Self {
+        Self::new_unchecked(
+            Uri::new_unchecked("https://blockprotocol.org/types/@blockprotocol/data-type/boolean"),
+            "Boolean",
+            "A True or False value".to_owned(),
+            "boolean",
+            HashMap::default(),
+        )
+    }
+
+    /// Returns the predefined null data type.
+    #[must_use]
+    pub fn null() -> Self {
+        Self::new_unchecked(
+            Uri::new_unchecked("https://blockprotocol.org/types/@blockprotocol/data-type/null"),
+            "Null",
+            "A placeholder value representing 'nothing'".to_owned(),
+            "null",
+            HashMap::default(),
+        )
+    }
+
+    /// Returns the predefined object data type.
+    #[must_use]
+    pub fn object() -> Self {
+        Self::new_unchecked(
+            Uri::new_unchecked("https://blockprotocol.org/types/@blockprotocol/data-type/object"),
+            "Object",
+            "A plain JSON object with no pre-defined structure".to_owned(),
+            "object",
+            HashMap::default(),
+        )
+    }
+
+    /// Returns the predefined empty-list data type.
+    #[must_use]
+    pub fn empty_list() -> Self {
+        Self::new_unchecked(
+            Uri::new_unchecked(
+                "https://blockprotocol.org/types/@blockprotocol/data-type/empty-list",
+            ),
+            "Empty List",
+            "An Empty List".to_owned(),
+            "array",
+            [("const".to_owned(), json!([]))].into_iter().collect(),
+        )
+    }
+}
+
+#[derive(Deserialize)]
+#[serde(rename_all = "camelCase")]
+enum DataTypeTag {
+    DataType,
+}
+
+#[derive(Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct DataTypeRepr {
```

Comment:
Can you explain why we have `Repr` and why we have so much duplicated code here. (Will probably ask for it to go into documentation but lets discuss in comment thread here)

Reply:
`Repr` is used for two reasons:
- It's not easy to implement a validator in serde, so we use a `TryInto` implementation to actually call `validate`, This way, every deserialization is validated, not only when creating them by hand. We currently don't have validation for `DataType` but we probably have later on. However, other structs like `PropertyTypeObject` has and this way it's very easy to add new validation later on without changing the surrounding code. This is the main reason we have `Repr`.
- The specs requires `"type" = "dataType"`. To avoid adding `DataTypeTag` on `DataType`, it's only used on `DataTypeRepr`.

Reply:
I simplified this where no validation currently happens: d2fbcfecbd8ae0b5235b5e3ffdc96688edbc11cc